### PR TITLE
Pipeline doesn't require any secrets

### PR DESCRIPTION
**Pull request by crazygao**
# Description
Current Replay Status
213 passed, 15 skipped,

Time cost about 1min 30s

cannot support the following
1. custom connect
2. python tool with complex returns
3. pf serve
4. Some record nits.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: .github/workflows/promptflow-replay-test.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    - name: Python Setup - ubuntu-latest - Python Version ${{ env.pyVersion }}
+      uses: "./.github/actions/step_create_python_environment"
+      with:
+        pythonVersion: ${{ env.pyVersion }}
+    - name: Build wheel
+      uses: "./.github/actions/step_sdk_setup"
+      with:
+        setupType: promptflow_with_extra
+        scriptPath: ${{ env.testWorkingDirectory }}
+    - name: Upload Wheel
```
Comment (wangchao1230) [inline .github/workflows/promptflow-replay-test.yml:37]: why do we need to upload wheel btw?

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py [hunk 2]
```diff
@@ -910,6 +879,7 @@ def check_connection_and_deployment(flow_folder, connection, deployment):
                 assert not (flow_folder / "azure_openai.yaml").exists()
                 assert not (flow_folder / "openai.yaml").exists()
 
+    @pytest.mark.skipif(RecordStorage.is_replaying_mode(), reason="cannot support interactive")
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:882]: btw, why cannot support interactive?

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_run.py [hunk 3]
```diff
@@ -821,6 +826,7 @@ def test_error_message_dump(self, pf):
         assert "error" in run_dict
         assert run_dict["error"] == exception
 
+    @pytest.mark.skipif(RecordStorage.is_replaying_mode(), reason="Skip in replaying mode")
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_run.py:829]: make the reason more detailed: Skip in replaying mode due to ...

File: src/promptflow/pyproject.toml [hunk 4]
```diff
@@ -5,6 +5,7 @@ line-length = 120
 markers = [
     "sdk_test",
     "cli_test",
+    "record_replay",
```
Comment (wangchao1230) [inline src/promptflow/pyproject.toml:8]: why do we need the marker?

File: src/promptflow/tests/sdk_cli_test/conftest.py [hunk 5]
```diff
@@ -182,25 +190,29 @@ def mock_invoke_tool(self, func, *args, **kwargs):
 
 
 @pytest.fixture
-def recording_enabled(mocker: MockerFixture):
+def recording_file_override(request: pytest.FixtureRequest, mocker: MockerFixture):
+    if RecordStorage.is_replaying_mode() or RecordStorage.is_recording_mode():
+        lowercase_cls_name = request.cls.__name__.lower()
+        file_path = RECORDINGS_TEST_CONFIGS_ROOT / f"{lowercase_cls_name}_node_cache.shelve"
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/conftest.py:196]: Can we use single binary file? Is separating to multilple files due to our tests could be running in parallel?
Add some comments on this

File: .github/workflows/promptflow-replay-test.yml (hunk 1 above)
Comment (crazygao) [reply .github/workflows/promptflow-replay-test.yml:37]: let the build dependencies doesn't affect the test environment.

I'll try this mode in this pipeline and will refactor all other pipelines.

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py (hunk 2 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:882]: It returns a stream object. Shelve record this and I'm not sure if I have
Skip for now and I'll investigate further

File: src/promptflow/pyproject.toml (hunk 4 above)
Comment (crazygao) [reply src/promptflow/pyproject.toml:8]: remove, marker for myself

File: src/promptflow/tests/sdk_cli_test/conftest.py (hunk 5 above)
Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/conftest.py:196]: let me investigate

File: .github/actions/step_publish_test_results/action.yml [hunk 6]
```diff
@@ -56,17 +56,17 @@ runs:
...
-  #       -H "X-GitHub-Api-Version: 2022-11-28" \
-  #       /repos/${{ github.repository }}/statuses/${{ github.sha }} \
-  #       -f state='success' \
-  #       -f target_url='https://github.com/microsoft/promptflow/actions/runs/${{ github.run_id }}' \
-  #       -f description='The build succeeded!' \
-  #       -f context='${{ inputs.context }}'
-  #   shell: bash -el {0}
-  #   env:
-  #     GITHUB_TOKEN: ${{ inputs.token }}
+  - name: Sending PR status to checkenforcer
```
Comment (zhengfeiwang) [inline .github/actions/step_publish_test_results/action.yml:59]: IIRC, `pull_request` cannot do this?

File: .github/workflows/promptflow-replay-test.yml [hunk 7]
```diff
@@ -0,0 +1,89 @@
+name: promptflow-replay-test
+on:
+  pull_request:
+    paths:
+      - src/promptflow/**
+      - scripts/**
```
Comment (zhengfeiwang) [inline .github/workflows/promptflow-replay-test.yml:6]: maybe we can have more strict on this condition?

File: .github/actions/step_publish_test_results/action.yml (hunk 6 above)
Comment (crazygao) [reply .github/actions/step_publish_test_results/action.yml:59]: let me remove this change, I copied this from original PR

File: src/promptflow/tests/sdk_cli_test/e2etests/test_executable.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+RUNS_DIR = "./tests/test_configs/runs"
+CONNECTIONS_DIR = "./tests/test_configs/connections"
+DATAS_DIR = "./tests/test_configs/datas"
+
+
+@pytest.mark.skipif(RecordStorage.is_replaying_mode(), reason="Missing executable deps in replay mode")
+@pytest.mark.usefixtures("use_secrets_config_file", "setup_local_connection", "install_custom_tool_pkg")
+@pytest.mark.cli_test
+@pytest.mark.e2etest
+class TestExecutable:
```
Comment (zhengfeiwang) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_executable.py:22]: is this test moved from other file? seems not related to replay test

Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_executable.py:22]: Yes, move out of test_cli, just a small refactor.

This test is skipped because it requires promptflow[executable]

File: src/promptflow/tests/sdk_cli_test/e2etests/test_flow_serve.py [hunk 9]
```diff
@@ -115,6 +119,7 @@ def test_chat_swagger(serving_client_llm_chat):
     }
 
 
+@pytest.mark.skipif(RecordStorage.is_replaying_mode(), reason="Doesn't support serve in replay mode")
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_flow_serve.py:122]: maybe add a class to group all these tests?

Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_flow_serve.py:122]: +1 on this, or open a task tracking this change; I think we should always use class for testing.

Comment (crazygao) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_flow_serve.py:122]: Create Task for skipped tests. cannot easy fix

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]
